File: app/src/main/java/com/deepanjanxyz/notepad/NoteEditorActivity.java
package com.deepanjanxyz.notepad;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Editor screen for a single note. Edits are auto-saved with a short debounce
 * and the note's database id is preserved across activity recreation so that
 * rotating or otherwise recreating the editor updates the existing row instead
 * of inserting a duplicate note.
 */
public class NoteEditorActivity extends AppCompatActivity {
    /** Auto-save debounce delay: at most one DB write per 500 ms instead of one per keystroke. */
    private static final long AUTOSAVE_DELAY_MS = 500L;
    /** Key used both for the intent extra and the saved instance state that carries the note's database id. */
    private static final String KEY_NOTE_ID = "note_id";

    private EditText etTitle, etContent;
    private DatabaseHelper dbHelper;
    private long noteId = -1;
    private FloatingActionButton fabSave;
    private Handler autosaveHandler;
    private boolean hasUnsavedChanges = false;

    /** Runs the pending auto-save when the debounce delay elapses. */
    private final Runnable saveRunnable = new Runnable() {
        @Override public void run() { saveNoteLocally(); }
    };

    /** Schedules a debounced auto-save whenever the title or content changes. */
    private final TextWatcher watcher = new TextWatcher() {
        @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
        @Override public void onTextChanged(CharSequence s, int start, int before, int count) { scheduleSave(); }
        @Override public void afterTextChanged(Editable s) {}
    };

    /** Sets up the editor: restores the note id after recreation, or loads the note from the launch intent. */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_note_editor);

        etTitle = findViewById(R.id.editTitle);
        etContent = findViewById(R.id.editContent);
        fabSave = findViewById(R.id.fabSave);
        dbHelper = new DatabaseHelper(this);
        autosaveHandler = new Handler(Looper.getMainLooper());

        if (savedInstanceState != null) {
            // Recreated after rotation/config change: keep the id of the note
            // this instance already saved so the next auto-save updates that
            // row instead of inserting a duplicate. The typed text is restored
            // automatically from the saved view state.
            noteId = savedInstanceState.getLong(KEY_NOTE_ID, -1);
        } else if (getIntent().hasExtra(KEY_NOTE_ID)) {
            noteId = getIntent().getLongExtra(KEY_NOTE_ID, -1);
            etTitle.setText(getIntent().getStringExtra("title"));
            etContent.setText(getIntent().getStringExtra("content"));
        }

        // Save button: save immediately and return to the notes list
        fabSave.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                saveNoteLocally();
                Toast.makeText(NoteEditorActivity.this, "Saved", Toast.LENGTH_SHORT).show();
                finish();
            }
        });

        // Debounced auto-save: text edits are flushed at most once every 500 ms
        etTitle.addTextChangedListener(watcher);
        etContent.addTextChangedListener(watcher);
    }

    /**
     * Persists the note's database id across activity recreation. Any pending
     * auto-save is flushed first so the saved id always matches the database.
     */
    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        autosaveHandler.removeCallbacks(saveRunnable);
        if (hasUnsavedChanges) saveNoteLocally();
        outState.putLong(KEY_NOTE_ID, noteId);
        super.onSaveInstanceState(outState);
    }

    /** (Re)schedules the debounced auto-save after a text change. */
    private void scheduleSave() {
        hasUnsavedChanges = true;
        autosaveHandler.removeCallbacks(saveRunnable);
        autosaveHandler.postDelayed(saveRunnable, AUTOSAVE_DELAY_MS);
    }

    /**
     * Writes the current title/content to the database: inserts a new row for a
     * brand-new note, updates the existing row afterwards, and deletes the
     * note entirely once all of its text has been cleared.
     */
    private void saveNoteLocally() {
        String title = etTitle.getText().toString();
        String content = etContent.getText().toString();
        String date = new SimpleDateFormat("MMM dd, HH:mm", Locale.getDefault()).format(new Date());
        hasUnsavedChanges = false;

        boolean hasText = !title.trim().isEmpty() || !content.trim().isEmpty();
        if (hasText) {
            if (noteId == -1) {
                noteId = dbHelper.insertNoteWithId(title, content, date);
            } else {
                dbHelper.updateNote(noteId, title, content, date);
            }
        } else if (noteId != -1) {
            // The note was emptied out: remove it instead of leaving a stale
            // copy of the previous text in the list
            dbHelper.deleteNote(noteId);
            noteId = -1;
        }
    }

    /** Flushes any pending auto-save so no typed text is ever lost when leaving the editor. */
    @Override
    protected void onPause() {
        super.onPause();
        autosaveHandler.removeCallbacks(saveRunnable);
        if (hasUnsavedChanges) saveNoteLocally();
    }

    /** Cancels any pending auto-save when the editor is destroyed. */
    @Override
    protected void onDestroy() {
        autosaveHandler.removeCallbacks(saveRunnable);
        super.onDestroy();
    }
}

File: app/src/main/java/com/deepanjanxyz/notepad/NoteEditorActivity.kt
package com.deepanjanxyz.notepad

import android.os.Bundle
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Done
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.activity.ComponentActivity
import kotlinx.coroutines.delay
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class NoteEditorActivity : ComponentActivity() {

    private val dbHelper by lazy { DatabaseHelper(this) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val initialNoteId = intent.getLongExtra(EXTRA_NOTE_ID, -1L)
        val initialTitle = intent.getStringExtra(EXTRA_TITLE) ?: ""
        val initialContent = intent.getStringExtra(EXTRA_CONTENT) ?: ""
        setContent {
            EditorScreen(initialNoteId, initialTitle, initialContent)
        }
    }

    @OptIn(ExperimentalMaterial3Api::class)
    @Composable
    private fun EditorScreen(initialNoteId: Long, initialTitle: String, initialContent: String) {
        var title by rememberSaveable { mutableStateOf(initialTitle) }
        var content by rememberSaveable { mutableStateOf(initialContent) }
        var noteId by rememberSaveable { mutableStateOf(initialNoteId) }

        // Last values actually written to the database. persist() compares
        // against these (not the launch-time values) so that reverting an
        // auto-saved edit is still detected as a change and gets saved.
        var lastPersistedTitle by rememberSaveable { mutableStateOf(initialTitle) }
        var lastPersistedContent by rememberSaveable { mutableStateOf(initialContent) }

        /**
         * Persists the note if anything actually changed since the last save.
         */
        fun persist() {
            if (title == lastPersistedTitle && content == lastPersistedContent) return
            val trimmedTitle = title.trim()
            val trimmedContent = content.trim()
            // Never persist a fully empty note (parity with the original app).
            if (trimmedTitle.isEmpty() && trimmedContent.isEmpty()) return
            val date = SimpleDateFormat(DATE_FORMAT, Locale.getDefault()).format(Date())
            if (noteId == -1L) {
                noteId = dbHelper.insertNote(trimmedTitle, trimmedContent, date)
            } else {
                dbHelper.updateNote(noteId, trimmedTitle, trimmedContent, date)
            }
            lastPersistedTitle = title
            lastPersistedContent = content
        }

        // Debounced auto-save on every change.
        LaunchedEffect(title, content) {
            delay(AUTO_SAVE_DELAY_MS)
            persist()
        }

        // Save one last time when the editor leaves the screen
        // (finish, configuration change, or process teardown).
        DisposableEffect(Unit) {
            onDispose { persist() }
        }

        fun saveAndFinish() {
            persist()
            Toast.makeText(this@NoteEditorActivity, R.string.note_saved, Toast.LENGTH_SHORT).show()
            finish()
        }

        Scaffold(
            topBar = {
                TopAppBar(
                    title = {
                        Text(
                            text = if (noteId == -1L) {
                                stringResource(R.string.new_note)
                            } else {
                                stringResource(R.string.edit_note)
                            },
                        )
                    },
                    navigationIcon = {
                        IconButton(onClick = { finish() }) {
                            Icon(
                                Icons.AutoMirrored.Filled.ArrowBack,
                                contentDescription = stringResource(R.string.action_back),
                            )
                        }
                    },
                    actions = {
                        IconButton(onClick = { saveAndFinish() }) {
                            Icon(
                                Icons.Default.Done,
                                contentDescription = stringResource(R.string.action_save),
                            )
                        }
                    },
                )
            },
        ) { innerPadding ->
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(innerPadding)
                    .imePadding(),
            ) {
                OutlinedTextField(
                    value = title,
                    onValueChange = { title = it },
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp, vertical = 8.dp),
                    placeholder = { Text(stringResource(R.string.note_title_hint)) },
                    singleLine = true,
                    textStyle = TextStyle(fontWeight = FontWeight.Bold, fontSize = 22.sp),
                )
                OutlinedTextField(
                    value = content,
                    onValueChange = { content = it },
                    modifier = Modifier
                        .fillMaxWidth()
                        .weight(1f)
                        .padding(horizontal = 16.dp)
                        .padding(bottom = 16.dp),
                    placeholder = { Text(stringResource(R.string.note_content_hint)) },
                )
            }
        }
    }

    companion object {
        const val EXTRA_NOTE_ID = "note_id"
        const val EXTRA_TITLE = "title"
        const val EXTRA_CONTENT = "content"
        private const val DATE_FORMAT = "MMM dd, HH:mm"
        private const val AUTO_SAVE_DELAY_MS = 400L
    }
}
